core/lib/multivm/src/versions/vm_latest/utils/fee.rs

```diff
@@ -32,18 +39,46 @@ pub(crate) fn derive_base_fee_and_gas_per_pubdata(
     )
 }
 
+pub(crate) fn get_batch_base_fee(l1_batch_env: &L1BatchEnv) -> u64 {
+    if let Some(base_fee) = l1_batch_env.enforced_base_fee {
+        return base_fee;
+    }
+    let (base_fee, _) =
+        derive_base_fee_and_gas_per_pubdata(l1_batch_env.fee_input.into_l1_pegged());
+    base_fee
+}
+
+pub(crate) fn get_batch_gas_per_pubdata(l1_batch_env: &L1BatchEnv) -> u64 {
+    derive_base_fee_and_gas_per_pubdata(l1_batch_env.fee_input.into_l1_pegged()).1
+}
+
 /// Changes the fee model output so that the expected gas per pubdata is smaller than or the `tx_gas_per_pubdata_limit`.
 /// This function expects that the currently expected gas per pubdata is greater than the `tx_gas_per_pubdata_limit`.
-pub(crate) fn adjust_l1_gas_price_for_tx(
-    fair_l2_gas_price: u64,
+pub(crate) fn adjust_pubdata_price_for_tx(
+    mut batch_fee_input: BatchFeeInput,
     tx_gas_per_pubdata_limit: U256,
-) -> u64 {
-    // gasPerPubdata = ceil(17 * l1gasprice / fair_l2_gas_price)
-    // gasPerPubdata <= 17 * l1gasprice / fair_l2_gas_price + 1
-    // fair_l2_gas_price(gasPerPubdata - 1) / 17 <= l1gasprice
-    let l1_gas_price = U256::from(fair_l2_gas_price)
-        * (tx_gas_per_pubdata_limit - U256::from(1u32))
-        / U256::from(17);
-
-    l1_gas_price.as_u64()
+) -> BatchFeeInput {
+    match &mut batch_fee_input {
+        BatchFeeInput::L1Pegged(fee_input) => {
+            // gasPerPubdata = ceil(17 * l1gasprice / fair_l2_gas_price)
+            // gasPerPubdata <= 17 * l1gasprice / fair_l2_gas_price + 1
+            // fair_l2_gas_price(gasPerPubdata - 1) / 17 <= l1gasprice
+            let new_l1_gas_price = U256::from(fee_input.fair_l2_gas_price)
+                * (tx_gas_per_pubdata_limit - U256::from(1u32))
+                / U256::from(17);
+
+            fee_input.l1_gas_price = new_l1_gas_price.as_u64();
+        }
+        BatchFeeInput::PubdataIndependent(fee_input) => {
+            // gasPerPubdata = ceil(fair_pubdata_price / fair_l2_gas_price)
+            // gasPerPubdata <= fair_pubdata_price / fair_l2_gas_price + 1
+            // fair_l2_gas_price(gasPerPubdata - 1) <= fair_pubdata_price
+            let new_fair_pubdata_price = U256::from(fee_input.fair_pubdata_price)
+                * (tx_gas_per_pubdata_limit - U256::from(1u32));
```

Comment:
If I understand correctly, then `fee_input.fair_l2_gas_price` should be used instead of `fee_input.fair_pubdata_price`

Reply:
good catch